pkg/telemetry/client.go

```diff
@@ -0,0 +1,241 @@
+// Copyright 2023 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// Package telemetry collects the CLI metrics and sends the telemetry data to supercollider
+package telemetry
+
+import (
+	"encoding/json"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+
+	"github.com/vmware-tanzu/tanzu-cli/pkg/buildinfo"
+	"github.com/vmware-tanzu/tanzu-cli/pkg/cli"
+	"github.com/vmware-tanzu/tanzu-cli/pkg/common"
+	configlib "github.com/vmware-tanzu/tanzu-plugin-runtime/config"
+	configtypes "github.com/vmware-tanzu/tanzu-plugin-runtime/config/types"
+)
+
+var once sync.Once
+
+var client MetricsHandler
+
+type MetricsHandler interface {
+	SetInstalledPlugins(plugins []cli.PluginInfo)
+	UpdateCmdPreRunMetrics(cmd *cobra.Command, args []string) error
+	UpdateCmdPostRunMetrics(metrics *PostRunMetrics) error
+	SaveMetrics() error
+	SendMetrics() error
+}
+
+type telemetryClient struct {
+	currentOperationMetrics *OperationMetricsPayload
+	installedPlugins        []cli.PluginInfo
+	metricsDB               MetricsDB
+}
+
+type PostRunMetrics struct {
+	ExitCode int
+}
+type OperationMetricsPayload struct {
+	CliID         string
+	StartTime     time.Time
+	EndTime       time.Time
+	Args          []string
+	NameArg       string
+	CommandName   string
+	ExitStatus    int
+	PluginName    string
+	Flags         string
+	CliVersion    string
+	PluginVersion string
+	Target        string
+	Endpoint      string
+	Error         string
+}
+
+func Client() MetricsHandler {
+	once.Do(func() {
+		client = NewTelemetryClient()
+	})
+	return client
+}
+
+func NewTelemetryClient() MetricsHandler {
```

Comment:
is there any specific reason we are making it public instead of private, as we have `Client()` to create a single instance of the client? 
If we make it public, we may end up creating multiple instances!!

Reply:
Good point. I will update the PR